### PR TITLE
feat(api-gateway): add debug mode

### PR DESCRIPTION
**Pull request by michaelbrewer**
**Issue #, if available:**

- https://github.com/awslabs/aws-lambda-powertools-roadmap/issues/4

Pairs with https://github.com/awslabs/aws-lambda-powertools-python/pull/506

## Description of changes:

Add debug mode support:

- Pretty print json request
- Pretty print json responses
- Output full stack trace

**Checklist**

 

* [X] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [X] Update tests
* [ ] Update docs
* [X] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws_lambda_powertools/event_handler/api_gateway.py [hunk 1]
```diff
@@ -237,20 +241,28 @@ def lambda_handler(event, context):
...
+    ):
         """
         Parameters
         ----------
         proxy_type: ProxyEventType
             Proxy request type, defaults to API Gateway V1
         cors: CORSConfig
             Optionally configure and enabled CORS. Not each route will need to have to cors=True
+        debug: Optional[bool]
+            Enables debug mode, by default False. Can be enabled by "POWERTOOLS_API_DEBUG" environment variable
```
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/api_gateway.py:258]: Could you change the env var to `POWERTOOLS_EVENT_HANDLER_DEBUG`?

This way we remain consistent with `POWERTOOLS_<utility>_<feature>`. This also means we could reuse in AppSync Event Handler or future ones

File: aws_lambda_powertools/shared/constants.py [hunk 2]
```diff
@@ -18,3 +18,5 @@
 
 XRAY_SDK_MODULE = "aws_xray_sdk"
 XRAY_SDK_CORE_MODULE = "aws_xray_sdk.core"
+
+API_DEBUG_ENV: str = "POWERTOOLS_API_DEBUG"
```
Comment (heitorlessa) [inline aws_lambda_powertools/shared/constants.py:22]:
```suggestion
EVENT_HANDLER_DEBUG_ENV: str = "POWERTOOLS_EVENT_HANDLER_DEBUG"
```

**Review by heitorlessa — CHANGES_REQUESTED** [review]
Looks good! One minor change in env var. QQ: how helpful would it be to literally `print` the event when this flag is enabled too?

I think this could be handled by Logger with the LOG_EVENT piece, so I'm on the fence.

**Comment by michaelbrewer** [comment]
> Looks good! One minor change in env var. QQ: how helpful would it be to literally `print` the event when this flag is enabled too?
> 
> 
> 
> I think this could be handled by Logger with the LOG_EVENT piece, so I'm on the fence.

Maybe as a debug tip to turn on log event environment variable.

**Comment by michaelbrewer** [comment]
@heitorlessa  ok it is all merged up.

**Comment by michaelbrewer** [comment]
@heitorlessa i added the print of the request

**Comment by michaelbrewer** [comment]
> Looks good! One minor change in env var. QQ: how helpful would it be to literally `print` the event when this flag is enabled too?
> 
> I think this could be handled by Logger with the LOG_EVENT piece, so I'm on the fence.

I added the print part

**Comment by michaelbrewer** [comment]
@heitorlessa  any other updates needed?

**Comment by heitorlessa** [comment]
Thanks a lot for the changes - I’ll check as soon as I can on Friday or
early next week

File: aws_lambda_powertools/shared/constants.py [hunk 3]
```diff
@@ -18,3 +18,5 @@
 
 XRAY_SDK_MODULE = "aws_xray_sdk"
 XRAY_SDK_CORE_MODULE = "aws_xray_sdk.core"
+
+EVENT_HANDLER_DEBUG_ENV: str = "POWERTOOLS_EVENT_HANDLER_DEBUG"
```
Comment (heitorlessa) [inline aws_lambda_powertools/shared/constants.py:22]: no need for :str as it's a constant

Comment (michaelbrewer) [reply aws_lambda_powertools/shared/constants.py:22]: sorry it is probably my over zealous ide

Comment (michaelbrewer) [reply aws_lambda_powertools/shared/constants.py:22]: in this context `XRAY_SDK_CORE_MODULE` is not consistent :P

File: tests/functional/event_handler/test_api_gateway.py [hunk 4]
```diff
@@ -582,3 +586,85 @@ def service_error():
...
+    mocked_print = MagicMock()
+    monkeypatch.setattr(builtins, "print", mocked_print)
+
+    # WHEN calling resolve
+    event = {"path": "/foo", "httpMethod": "GET"}
+    app(event, None)
+
+    # THEN print the event
+    # NOTE: other calls might have happened outside of this mock
+    mocked_print.assert_any_call(json.dumps(event, indent=4))
```
Comment (heitorlessa) [inline tests/functional/event_handler/test_api_gateway.py:670]: You can use Pytest `capsys` fixture to capture stdout that `print` builtin used.

```suggestion
def test_debug_print_event(capsys):
    # GIVE debug is True
    app = ApiGatewayResolver(debug=True)

    # WHEN calling resolve
    event = {"path": "/foo", "httpMethod": "GET"}
    app(event, None)

    # THEN print the event
    out, err = capsys.readouterr()
    assert json.loads(out) == event
```

**Review by heitorlessa — CHANGES_REQUESTED** [review]
That's great, one tiny change so we don't use mock in a functional test and capture stdout instead, then it's good to merge :) 

You know, it'd be great to have permissions to push these so I don't have to bother ya ;)

File: aws_lambda_powertools/shared/constants.py (hunk 3 above)
Comment (heitorlessa) [reply aws_lambda_powertools/shared/constants.py]: haha I said it, then I went over to look... and I did the same in others except XRAY, then thought FML... that needs fixing

**Review by heitorlessa — APPROVED** [review]

**Comment by michaelbrewer** [comment]
> That's great, one tiny change so we don't use mock in a functional test and capture stdout instead, then it's good to merge :)
> 
> You know, it'd be great to have permissions to push these so I don't have to bother ya ;)

i think that only works if i do it via my personal account :)

**Pull request merged**